### PR TITLE
[DOC] Documentation for Bidirectional FM index

### PR DESCRIPTION
**Pull request by cpockrandt**

### CODE REVIEW

File: include/seqan/index/index_fm_rank_dictionary_levels.h [hunk 1]
```diff
@@ -124,10 +124,41 @@ struct RankDictionaryEntry_;
...
+ *
+ * @brief LevelsPrefixRDConfig allows configuring a @link Levels @endlink RankDictionary that is recommended for fast searching in bidirectional indices.
+ *
+ * @signature template <typename TSize = size_t, typename TFibre = Alloc<>, unsigned LEVELS_ = 1, unsigned WORDS_PER_BLOCK_ = 0>
+ *            struct LevelsRDConfig<TSize, TFibre, LEVELS_, WORDS_PER_BLOCK_>;
+ *
+ * @tparam TSize           A data type that can store the length of the input text. Default: <tt>size_t</tt>
+ * @tparam TFibre          A tag for specialization purposes of the underlying strings. Default: <tt>Alloc<></tt>
+ * @tparam LEVELS          The number of levels (1, 2, or 3). The more levels, the lower the space consumption but possibly slight performance decreases. Default: <tt>1</tt>
+ * @tparam WORDS_PER_BLOCK The number of popcount operations per rank query. A lower number implies more space for faster runtime. 0 is a shortcut for the size of the alphabet of the RankDictionary. Default: <tt>0</tt>
```
Comment (rrahn) [inline include/seqan/index/index_fm_rank_dictionary_levels.h]: Same here!

File: include/seqan/index/index_fm_rank_dictionary_wt.h [hunk 2]
```diff
@@ -48,8 +48,25 @@ namespace seqan {
...
+ * @brief WTRDConfig allows configuring a @link WaveletTree @endlink.
+ *
+ * @signature template <typename TSize = size_t, typename TFibre = Alloc<>, unsigned LEVELS_ = 1, unsigned WORDS_PER_BLOCK_ = 0>
+ *            struct WTRDConfig<TSize, TFibre, LEVELS_, WORDS_PER_BLOCK_>;
+ *
+ * @tparam TSize           A data type that can store the length of the input text. Default: <tt>size_t</tt>
+ * @tparam TFibre          A tag for specialization purposes of the underlying strings. Default: <tt>Alloc<></tt>
+ * @tparam LEVELS          The number of levels (1, 2, or 3). The more levels, the lower the space consumption but possibly slight performance decreases. Default: <tt>1</tt>
+ * @tparam WORDS_PER_BLOCK The number of popcount operations per rank query. A lower number implies more space for faster runtime. 0 is a shortcut for the size of the alphabet of the RankDictionary. Default: <tt>0</tt>
+ *
```
Comment (rrahn) [inline include/seqan/index/index_fm_rank_dictionary_wt.h]: Remove empty line and unify template names.

File: include/seqan/index/index_fm_rank_dictionary_wt.h [hunk 3]
```diff
@@ -48,8 +48,25 @@ namespace seqan {
...
+ */
+
+template <typename TSize = size_t, typename TFibre = Alloc<>, unsigned LEVELS = 1, unsigned WORDS_PER_BLOCK = 0/*, unsigned ARITY_ = 2*/>
+struct WTRDConfig : LevelsRDConfig<TSize, TFibre, LEVELS, WORDS_PER_BLOCK>
```
Comment (rrahn) [inline include/seqan/index/index_fm_rank_dictionary_wt.h]: For some structs use the underscore notation and for some you don't. Could you unify this, or does it have a special meaning?

File: manual/source/Tutorial/DataStructures/Indices/FMIndex.rst [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+.. includefrags:: demos/tutorial/indices/fm_index.cpp
+   :fragment: FMIndexConfigs
+
+Optimizing Space or Running Time
+--------------------------------
+
+All three rank dictionaries are based on bit vectors with constant-time rank support. We skip at this point details on how rank queries work and focus
+on the parameters and their effects. To reduce the space consumption there are three parameters that you can adjust:
+
+First of all TSize should be the smallest data type that can store the length of the indexed text (e.g. uint8_t, uint16_t, uint32_t, uint64_t).
```
Comment (rrahn) [inline manual/source/Tutorial/DataStructures/Indices/FMIndex.rst]: emphasis TSize, as well as uint_8 etc.

File: manual/source/Tutorial/DataStructures/Indices/FMIndex.rst [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+For a :dox:`StringSet` the length is defined as the the sum of lengths of strings plus the number of strings.
+
+The bit vector is clustered into blocks that store the precomputed rank up to every block. The reduce the space of these blocks, one can add an additional level of
```
Comment (rrahn) [inline manual/source/Tutorial/DataStructures/Indices/FMIndex.rst]: To reduce the ...

File: manual/source/Tutorial/DataStructures/Indices/FMIndex.rst [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+Bidirectional FMIndex
+---------------------
+
+The bidirectional FMIndex can be used to search a pattern into both directions, i.e. a string can be extend by a character to the left or right in an arbitrary manner. For information on how
+to search in a bidirectional FMIndex, please check the Tutorial on :ref:`tutorial-datastructures-indices-index-iterators`.
+
+.. includefrags:: demos/tutorial/indices/fm_index.cpp
+   :fragment: BidirectionalIndex
+
+TODO: goDown(it, Rev()), goDown(it, Fwd()), goRoot, roRight?, goUp, getOccurrences, getRepresentative. Please check the API dox for more details.
```
Comment (rrahn) [inline manual/source/Tutorial/DataStructures/Indices/FMIndex.rst]: Can you hide the todo from the documentation page, please.

File: manual/source/Tutorial/DataStructures/Indices/IndexIterators.rst [hunk 7]
```diff
@@ -168,10 +168,44 @@ Assignment 3
...
+trie). Using the **Rev** tag accesses the reversed text iterator and extends the pattern to the right.
+
+Creating the index and iterator is very similar to unidirectional indices. The FMIndex is wrapped in a
+BidirectionalIndex tag:
+
+.. includefrags:: demos/tutorial/index_iterators/index_bidirectional_search.cpp
+   :fragment: Create
+
+All methods for traversing the virtual trie are extended by the direction tag **Fwd** or **Rev**. If none is used, it
+will access the iterator on the original text by default (same as using the **Fwd** tag). The goUp method is the only
```
Comment (rrahn) [inline manual/source/Tutorial/DataStructures/Indices/IndexIterators.rst]: Emphasis goUp

File: manual/source/Tutorial/DataStructures/Indices/IndexIterators.rst [hunk 8]
```diff
@@ -168,10 +168,44 @@ Assignment 3
...
+method that does not specify a direction tag. goUp corresponds to an undo operation, i.e. it rolls both iterators back
```
Comment (rrahn) [inline manual/source/Tutorial/DataStructures/Indices/IndexIterators.rst]: emphasis goUp

**Comment by cpockrandt** [comment]
@rrahn Thanks for the hints :grin:

**Pull request merged**